### PR TITLE
Added in support for custom debug logging

### PR DESCRIPTION
**Pull request by niftymonkey**
Added a property to the Reactor constructor object that allows you to pass in your own custom logging implementation.  Tested using the following (in the flux-chat example):
```
const customLogger = {
  dispatchStart: function(reactorState, type, payload) {
    console.log('*************************************')
    console.log('     TYPE: ', type)
    console.log('  PAYLOAD: ', payload)
  },
  dispatchError: function(reactorState, error) {
    console.error(error, reactorState.state.toJS())
  },
  dispatchEnd: function(reactorState, state, dirtyStores) {
    console.log('OLD STATE: ', reactorState.state.toJS())
    console.log('  CHANGED: ', dirtyStores.toList().toJS())
    console.log('NEW STATE: ', state.toJS())
  },
}

module.exports = new Nuclear.Reactor({
  debug: true,
  logger: customLogger,
})
```

### CODE REVIEW

**Comment by jordangarcia** [comment]
This looks awesome, will do a more thorough review this afternoon

**Comment by jordangarcia** [comment]
@niftymonkey I'd like to merge this in but wait a bit before a major version bumb is released.

Would it be okay if this got merged and tagged with some `beta` tag that you could pull in via a github reference in your `package.json`?

**Comment by niftymonkey** [comment]
Yeah, that sounds fine.  Thanks a bunch!

File: src/reactor.js [hunk 1]
```diff
@@ -26,9 +27,13 @@ class Reactor {
   constructor(config = {}) {
     const debug = !!config.debug
     const baseOptions = debug ? DEBUG_OPTIONS : PROD_OPTIONS
+    // if defined, merge the custom implementation over the noop logger to avoid undefined lookups,
+    // otherwise, just use the built-in console group logger
+    const debugLogger = config.logging ? extend({}, NoopLogger, config.logging) : ConsoleGroupLogger
```
Comment (jordangarcia) [inline src/reactor.js]: Since we are passing `Loggers` here i think the config option should be called `Logger`

**Comment by jordangarcia** [comment]
Checked the code out and played with it, looks really great.  Thanks for doing.

One minor change, then LGTM

File: src/reactor.js (hunk 1 above)
Comment (niftymonkey) [inline src/reactor.js]: Sure thing, did you want it to be uppercase like that?  Or did you just want me to change `logging` to `logger`?

Comment (niftymonkey) [inline src/reactor.js]: Changed to `logger` for consistency with the rest of the options.  Let me know if you want it changed to first-cap.

File: src/reactor.js [hunk 2]
```diff
@@ -26,9 +27,13 @@ class Reactor {
   constructor(config = {}) {
     const debug = !!config.debug
     const baseOptions = debug ? DEBUG_OPTIONS : PROD_OPTIONS
+    // if defined, merge the custom implementation over the noop logger to avoid undefined lookups,
+    // otherwise, just use the built-in console group logger
+    const debugLogger = config.logger ? extend({}, NoopLogger, config.logger) : ConsoleGroupLogger
     const initialReactorState = new ReactorState({
       debug: debug,
       cache: config.cache || DefaultCache(),
+      logger: debug ? debugLogger : NoopLogger,
```
Comment (jordangarcia) [inline src/reactor.js]: The way the current code is written it's not possible to have a logger when `debug` is falsy.

It'd be nice to be able to do:

```js
new Nuclear.Reactor({
  logger: MyLogger
})
```

A little context, the `debug` option is very old and existing before specific logging options were available, so I kept it around to maintain backwards compatibility.

Ideally, regardless of the value of `debug` if you pass in a `logger` option it should respect that.

```js
let logger = config.logger ? extend({}, NoopLogger, config.logger) : NoopLogger
if (!config.logger && debug) {
  logger = ConsoleGroupLogger
}
```

I believe that logic preserves the use cases I mentioned above.

**Comment by jordangarcia** [comment]
One last comment regarding respecting the `logger` option regardless of `debug` :)

File: src/reactor.js (hunk 2 above)
Comment (niftymonkey) [inline src/reactor.js]: Will do! 👍

Comment (niftymonkey) [inline src/reactor.js]: Ok, got that checked in.

**Comment by jordangarcia** [comment]
@niftymonkey thanks so much for adding this functionality.

I'll merge in now and cut a new release this afternoon

**Comment by niftymonkey** [comment]
Seriously appreciate your quick turnaround on this one.  Thanks for that!

**Comment by niftymonkey** [comment]
@jordangarcia - Were you thinking the `beta` tag was what you were wanting to go with, or should I go with a commit-ish git url?

**Comment by jordangarcia** [comment]
I'm going to a 1.5 release tomorrow with the logging changes. Using the
commit hash right now won't work because the dust files haven't been built

**Comment by jordangarcia** [comment]
S/bust/dist

**Pull request merged**